src/cartesian_path.cpp

```diff
@@ -0,0 +1,111 @@
+INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+/* Author: Karl Kruusamäe */
+/* E-mail: karl.kruusamae@ut.ee */
+
+#include <ros/ros.h>
+#include <moveit/move_group_interface/move_group_interface.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <geometry_msgs/Pose.h>
+#include <vector>
+
+int main(int argc, char** argv)
+{
+  // Basic ROS setup
+  // ^^^^^^^^^^^^^^^
+  ros::init(argc, argv, "cartesian_path");
+  ros::NodeHandle node_handle;
+  ros::AsyncSpinner spinner(1);
+  spinner.start();
+
+  // MoveGroupInterface API can be found at:
+  // http://docs.ros.org/kinetic/api/moveit_ros_planning_interface/html/classmoveit_1_1planning__interface_1_1MoveGroupInterface.html
+
+  // MoveIt setup
+  // ^^^^^^^^^^^^
+  // A MoveGroupInterface instance (a client for the MoveGroup action) can be easily setup using just the name of the
+  // planning group you would like to control and plan for.
+  moveit::planning_interface::MoveGroupInterface move_group("xarm7");
+
+  // Planning a Cartesian path
+  // ^^^^^^^^^^^^^^^^^^^^^^^^^
+  // We can plan a Cartesian path by specifying a series of waypoints for the end-effector to go through.
+  // As each waypoint will be defined as a pose, which requires 7 variables - x, y, and z for the position and x, y, z,
+  // and w for the orientation - to be defined, it is convenient to first get the current end-effector pose and then
+  // modify it to a suitable waypoint.
+
+  // Getting the current pose of the end-effector.
+  geometry_msgs::PoseStamped current_pose;
+  current_pose = move_group.getCurrentPose();
+
+  // Creating a vector that contains all the waypoints of the desired Cartesian path.
+  std::vector<geometry_msgs::Pose> waypoints;
+
+  // Modifying the current pose into a next waypoint and then each waypoint into a subsequent one.
+  // Feel free to change the following position values to see their influence on the motion planning.
+  geometry_msgs::Pose next_waypoint = current_pose.pose;
+
+  // Move diagonally in xy-plane
+  next_waypoint.position.x += 0.2;
+  next_waypoint.position.y -= 0.2;
+  waypoints.push_back(next_waypoint);
+
+  // Move only along y-axis
+  next_waypoint.position.y += 0.2;
+  waypoints.push_back(next_waypoint);
+
+  // Move diagonally in xy-plane
+  next_waypoint.position.x -= 0.2;
+  next_waypoint.position.y += 0.2;
+  waypoints.push_back(next_waypoint);
+
```

Comment:
I suggest planning a closed-loop trajectory, so that when executing repetitively we stay inside arm's workspace.
```suggestion
  // Move only along y-axis to get back to where we started.
  next_waypoint.position.y -= 0.2;
  waypoints.push_back(next_waypoint);
```

Reply:
i actually gave a lot of thought about which trajectory/waypoints to demonstrate in this example. here are some of the considerations that made me choose this in-plane non-closed-loop trajectory:
1) if the demo is in one plane, then the follow-up task during training/lab can be for the learners to add movement in z-direction;
2) if the demo is not closed-loop, it is possible to run the same node over and over for 2-3 runs to reach a state where fraction is less than zero. the meaning of fraction is a learning point;
3) closing the loop can be a follow-up task for the learners;
4) the demo is not showing the alteration of the orientation of the eef to make code easier to read and adding orientations can be an advanced level follow-up task for learners.

@Veix123 @RValner what are your thoughts about the above considerations? objections? suggestions?

Reply:
Great ideas! I still see the basic template as a "stable" and repetitive program. All the tasks @kruusamae mentioned above could be added as additional assignments on top of that teaching different aspects of the interface.

Reply:
> I still see the basic template as a "stable" and repetitive program.
what about the pose_goal node then? no closed loop behaviour there. if the learner runs the same node multiple times, sooner or later the motion planning will fail ;)

Reply:
@RValner weigh in, please! ;)
@Veix123 any chance you've had a change of heart during the past 1 month or so? :)

Reply:
> what about the pose_goal node then? no closed loop behaviour there.
Feelings still mixed ;) Considering "out of workspace" planning failures as important learning aspects, I do see that the overall approach would be more systematic if you ignore my suggestion above.

Reply:
I'm completely ok with the idea of the students encountering planning failure issues rather sooner than later + code does look simpler.

Reply:
as @Veix123 is still requesting changes, i can't do a _clean_ merge. i feel we can close this discussion by keeping the code as is. maybe in the future, after we've done more trainings with this package, we'll revise some of the concepts.

Reply:
I agree, approved for now!